### PR TITLE
[libstd]: add Dir.Iterator tests

### PR DESCRIPTION
**Pull request by kubkon**
This commit adds some `std.fs.Dir.Iterator` tests.

Partially addresses #5653.

### CODE REVIEW

**Comment by kubkon** [comment]
SourceHut config seems broken -- see #5756.

**Comment by squeek502** [comment]
#5653 is proving its worth with every PR. :slightly_smiling_face: Another bug uncovered.

Windows is getting:
```
Entry{ .name = some_fil, .kind = Kind.Directory }
Entry{ .name = some_file, .kind = Kind.File }
```

**Comment by kubkon** [comment]
> #5653 is proving its worth with every PR. 🙂 Looks like this test is uncovering another bug.
> 
> Windows is getting:
> 
> ```
> Entry{ .name = some_fil, .kind = Kind.Directory }
> Entry{ .name = some_file, .kind = Kind.File }
> ```
> 
> EDIT: Within the loop, the entry is correct, it's only after it's appended to entries that it becomes corrupted. Looks like the `name` slice is being reused.
> 
> EDIT#2: Maybe not a bug, this is the doc comment for `Dir.Iterator.next`:
> 
> ```
> /// Memory such as file names referenced in this returned entry becomes invalid
> /// with subsequent calls to `next`, as well as when this `Dir` is deinitialized.
> ```
> 
> EDIT#3: Alternate implementation that fixes the test on Windows:
> 
> ```zig
> test "Dir.Iterator" {
>     var tmp_dir = tmpDir(.{ .iterate = true });
>     defer tmp_dir.cleanup();
> 
>     // First, create a couple of entries to iterate over.
>     const file = try tmp_dir.dir.createFile("some_file", .{});
>     file.close();
> 
>     try tmp_dir.dir.makeDir("some_dir");
> 
>     var iter = tmp_dir.dir.iterate();
>     var len: usize = 0;
> 
>     while (try iter.next()) |entry| {
>         switch (entry.kind) {
>             .File => testing.expectEqualSlices(u8, "some_file", entry.name),
>             .Directory => testing.expectEqualSlices(u8, "some_dir", entry.name),
>             else => unreachable,
>         }
>         len += 1;
>     }
> 
>     testing.expect(len == 2); // note that the Iterator skips '.' and '..'
> }
> ```
> 
> EDIT#4: Also worth noting that the doc comment for `Dir.Iterator.next` is missing in the Windows implementation, which should also be fixed.

Wow, thanks for the in-depth analysis @squeek502!

I've actually had the same observations just didn't get a chance to write that up yet. Indeed, `name` is reused per `next` call, however, this doesn't change the fact we have an inconsistency between hosts, so a doc comment specifically outlining this on Windows should be added.

Now, this still doesn't change the fact we should be able to make the test pass without resorting to testing for equality within the `while` loop. I reckon we should be able to copy out the bytes into an `ArrayList` (i.e., collect the entries) and continue on with the test as is, right? In fact, I assumed that `ArrayList.append` would do the copying for us since according to the docs:

> /// Extend the list by 1 element. Allocates more memory as necessary.

It's obviously not the case. Anyhow, I tried copying the bytes out manually yesterday, and I'd still get memory corruption somewhere for some reason. It was late though, so maybe I'll have better luck at some point later today. If anything comes back, I shall drop a line or two here.

**Comment by kubkon** [comment]
OK, 9b29218 should fix the memory corruption bug in the test. While we're here, do you think it would be useful to have `Iterator.nextAlloc(self: *Self, allocator: *Allocator)` method which internally would call `next()`, and allocate and copy the entry into the newly allocated space?

File: lib/std/fs/test.zig [hunk 1]
```diff
@@ -3,10 +3,53 @@ const testing = std.testing;
...
+    defer arena.deinit();
+
+    var entries = std.ArrayList(Dir.Entry).init(&arena.allocator);
+
+    // Create iterator.
+    var iter = tmp_dir.dir.iterate();
+    while (try iter.next()) |entry| {
+        // We cannot just store `entry` as on Windows, we're re-using the name buffer
+        // which means we'll actually share the `name` pointer between entries!
+        const name = try mem.dupe(&arena.allocator, u8, entry.name);
```
Comment (joachimschmidt557) [inline lib/std/fs/test.zig:33]:
```suggestion
        const name = try arena.allocator.dupe(u8, entry.name);
```
`std.mem.dupe` is deprecated now

Comment (kubkon) [reply lib/std/fs/test.zig:33]: Great catch, thanks!

**Comment by kubkon** [comment]
Unless anyone else has anything else to add, could we merge this into upstream?

**Comment by squeek502** [comment]
> Indeed, name is reused per next call, however, this doesn't change the fact we have an inconsistency between hosts, so a doc comment specifically outlining this on Windows should be added.

Might be worth addressing this in this PR.

**Comment by kubkon** [comment]
> > Indeed, name is reused per next call, however, this doesn't change the fact we have an inconsistency between hosts, so a doc comment specifically outlining this on Windows should be added.
> 
> Might be worth addressing this in this PR.

Good catch, thanks! Done in 417c928.

**Pull request merged**